File: src/vec3.zig
const std = @import("std");

pub const Vec3 = struct {
    e: [3]f64,

    //
    // Constructors
    //

    pub fn init(e0: f64, e1: f64, e2: f64) Vec3 {
        return Vec3{ .e = .{ e0, e1, e2 } };
    }

    pub fn zero() Vec3 {
        return Vec3{ .e = .{ 0.0, 0.0, 0.0 } };
    }

    //
    // Modify
    //

    pub fn add(self: Vec3, other: Vec3) Vec3 {
        return Vec3{.e{
            self.e[0] + other.e[0],
            self.e[1] + other.e[1],
            self.e[2] + other.e[2],
        }};
    }

    pub fn sub(self: Vec3, other: Vec3) Vec3 {
        return Vec3{.e{
            self.e[0] - other.e[0],
            self.e[1] - other.e[1],
            self.e[2] - other.e[2],
        }};
    }

    pub fn mult_scalar(self: Vec3, t: f64) Vec3 {
        return Vec3{.e{
            self.e[0] * t,
            self.e[1] * t,
            self.e[2] * t,
        }};
    }

    pub fn div_scalar(self: Vec3, t: f64) Vec3 {
        return self.mult_scalar(1.0 / t);
    }

    pub fn length(self: Vec3) f64 {
        return std.math.sqrt(self.length_squared());
    }

    pub fn length_squared(self: Vec3) f64 {
        return self.e[0] * self.e[0] + self.e[1] * self.e[1] + self.e[2] * self.e[2];
    }

    pub fn unit_vector(self: Vec3) Vec3 {
        return self.div_scalar(self.length());
    }
};

// Semantic aliases
pub const Point3 = Vec3;

// Unary subtraction
pub fn neg(v: Vec3) Vec3 {
    return Vec3{
        .e {
            -v.e[0],
            -v.e[1],
            -v.e[2],
        }
    };
}

//
// Binary operations
//

pub fn add(u: Vec3, v: Vec3) Vec3 {
    return u.add(v);
}

pub fn sub(u: Vec3, v: Vec3) Vec3 {
    return u.sub(v);
}

pub fn mult(u: Vec3, v: Vec3) Vec3 {
    return Vec3 {
        .e {
            u.e[0] * v.e[0],
            u.e[1] * v.e[1],
            u.e[2] * v.e[2],
        }
    };
}

pub fn mult_scalar(t: f64, v: Vec3) Vec3 {
    return v.mult_scalar(t);
}

pub fn mult_vec3_scalar(v: Vec3, t: f64) Vec3 {
    return mult_scalar(t, v);
}

pub fn div_scalar(v: Vec3, t: f64) Vec3 {
    return v.div_scalar(t);
}

// Dot product
pub fn dot(u: Vec3, v: Vec3) f64 {
    return u.e[0] * v.e[0] + u.e[1] * v.e[1] + u.e[2] * v.e[2];
}

// Cross product
pub fn cross(u: Vec3, v: Vec3) Vec3 {
    return Vec3{
        .e = .{
            u.e[1] * v.e[2] - u.e[2] * v.e[1],
            u.e[2] * v.e[0] - u.e[0] * v.e[2],
            u.e[0] * v.e[1] - u.e[1] * v.e[0],
        },
    };
}

// Print the mf
//
// Pass some sort of stdout here, anything that can have a print function
pub fn print_vec3(w: anytype, v: Vec3) !void {
    try w.print("{d} {d} {d}\n", .{v.e[0], v.e[1], v.e[2]});
}

File: src/vec3.rs
use core::fmt;
use std::ops::{Add, AddAssign, Div, DivAssign, Index, IndexMut, Mul, MulAssign, Neg, Sub};

#[derive(Default, Copy, Clone)]
pub struct Vec3 {
    e: [f64; 3],
}

impl Vec3 {
    //
    // Struct lifecycle.
    //

    pub fn new(e0: f64, e1: f64, e2: f64) -> Self {
        Self { e: [e0, e1, e2] }
    }

    pub fn zero() -> Self {
        Self::default()
    }

    //
    // Accessors -> gay kkkk.
    //

    pub fn x(&self) -> f64 {
        self.e[0]
    }

    pub fn y(&self) -> f64 {
        self.e[1]
    }

    pub fn z(&self) -> f64 {
        self.e[2]
    }

    //
    // Length methods.
    //

    pub fn length(&self) -> f64 {
        self.length_squared().sqrt()
    }

    pub fn length_squared(&self) -> f64 {
        self.e[0] * self.e[0] + self.e[1] * self.e[1] + self.e[2] * self.e[2]
    }

    //
    // Mutating operators.
    //

    pub fn add_assign(&mut self, other: Vec3) {
        self.e[0] += other.e[0];
        self.e[1] += other.e[1];
        self.e[2] += other.e[2];
    }

    pub fn mul_assign(&mut self, t: f64) {
        self.e[0] *= t;
        self.e[1] *= t;
        self.e[2] *= t;
    }

    pub fn div_assign(&mut self, t: f64) {
        let inv_t = 1.0 / t;
        self.mul_assign(inv_t);
    }

    //
    // Utils.
    //

    pub fn dot(&self, other: &Vec3) -> f64 {
        self.e[0] * other.e[0] + self.e[1] * other.e[1] + self.e[2] * other.e[2]
    }

    pub fn cross(&self, other: &Vec3) -> Vec3 {
        Vec3 {
            e: [
                self.e[1] * other.e[2] - self.e[2] * other.e[1],
                self.e[2] * other.e[0] - self.e[0] * other.e[2],
                self.e[0] * other.e[1] - self.e[1] * other.e[0],
            ],
        }
    }

    pub fn unit_vector(&self) -> Vec3 {
        *self / self.length()
    }
}

//
// Operator overloading.
//

impl Neg for Vec3 {
    type Output = Vec3;

    fn neg(self) -> Self::Output {
        Vec3::new(-self.e[0], -self.e[1], -self.e[2])
    }
}

/// Indexing.
impl Index<usize> for Vec3 {
    type Output = f64;

    fn index(&self, idx: usize) -> &f64 {
        &self.e[idx]
    }
}

/// Mut index.
impl IndexMut<usize> for Vec3 {
    fn index_mut(&mut self, idx: usize) -> &mut f64 {
        &mut self.e[idx]
    }
}

// Addition
impl Add for Vec3 {
    type Output = Vec3;

    fn add(self, other: Vec3) -> Vec3 {
        Vec3::new(
            self.e[0] + other.e[0],
            self.e[1] + other.e[1],
            self.e[2] + other.e[2],
        )
    }
}

impl AddAssign for Vec3 {
    fn add_assign(&mut self, other: Vec3) {
        self.e[0] += other.e[0];
        self.e[1] += other.e[1];
        self.e[2] += other.e[2];
    }
}

// Subtraction.
impl Sub for Vec3 {
    type Output = Vec3;

    fn sub(self, other: Vec3) -> Vec3 {
        Vec3::new(
            self.e[0] - other.e[0],
            self.e[1] - other.e[1],
            self.e[2] - other.e[2],
        )
    }
}

// Scalar multiplication
impl Mul<f64> for Vec3 {
    type Output = Vec3;

    fn mul(self, t: f64) -> Vec3 {
        Vec3::new(self.e[0] * t, self.e[1] * t, self.e[2] * t)
    }
}

impl Mul<Vec3> for f64 {
    type Output = Vec3;

    fn mul(self, v: Vec3) -> Vec3 {
        v * self
    }
}

impl MulAssign<f64> for Vec3 {
    fn mul_assign(&mut self, t: f64) {
        self.e[0] *= t;
        self.e[1] *= t;
        self.e[2] *= t;
    }
}

// Scalar division
impl Div<f64> for Vec3 {
    type Output = Vec3;

    fn div(self, t: f64) -> Vec3 {
        let inv_t = 1.0 / t;
        self * inv_t
    }
}

impl DivAssign<f64> for Vec3 {
    fn div_assign(&mut self, t: f64) {
        let inv_t = 1.0 / t;
        self.e[0] *= inv_t;
        self.e[1] *= inv_t;
        self.e[2] *= inv_t;
    }
}

// Component-wise multiplication
impl Mul for Vec3 {
    type Output = Vec3;

    fn mul(self, other: Vec3) -> Vec3 {
        Vec3::new(
            self.e[0] * other.e[0],
            self.e[1] * other.e[1],
            self.e[2] * other.e[2],
        )
    }
}

// Display formatting
impl fmt::Display for Vec3 {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} {} {}", self.e[0], self.e[1], self.e[2])
    }
}

// Type alias for clarity
pub type Point3 = Vec3;

// Utility functions
pub fn dot(u: &Vec3, v: &Vec3) -> f64 {
    u.dot(v)
}

pub fn cross(u: &Vec3, v: &Vec3) -> Vec3 {
    u.cross(v)
}

pub fn unit_vector(v: &Vec3) -> Vec3 {
    v.unit_vector()
}
